ui/apps/platform/src/Containers/Dashboard/PatternFly/hooks/useViolationCounts.ts

```diff
@@ -0,0 +1,64 @@
+import { useEffect, useState } from 'react';
+import { sortBy } from 'lodash';
+
+import {
+    AlertGroup,
+    AlertQueryGroupBy,
+    fetchSummaryAlertCounts,
+    Severity,
+} from 'services/AlertsService';
+
+function pluckSeverityCount(severity: Severity): (group: AlertGroup) => number {
+    return ({ counts }) => {
+        const severityCount = counts.find((ct) => ct.severity === severity)?.count || '0';
+        return -parseInt(severityCount, 10);
+    };
+}
+
+export type UseViolationCountsReturn = {
+    violationCounts: AlertGroup[];
+    loading: boolean;
+    error: Error | null;
+};
+
+export default function useViolationCounts(
+    groupBy: AlertQueryGroupBy,
+    query: string,
+    limit = 5
+): UseViolationCountsReturn {
+    const [violationCounts, setViolationCounts] = useState<AlertGroup[]>([]);
+    const [loading, setLoading] = useState<boolean>(true);
+    const [error, setError] = useState<Error | null>(null);
+
+    useEffect(() => {
+        const { request, cancel } = fetchSummaryAlertCounts({
+            'request.query': query,
+            group_by: groupBy,
+        });
+        request
+            .then((groups) => {
+                const selected = sortBy(groups, [
+                    pluckSeverityCount('CRITICAL_SEVERITY'),
+                    pluckSeverityCount('HIGH_SEVERITY'),
+                    pluckSeverityCount('MEDIUM_SEVERITY'),
+                    pluckSeverityCount('LOW_SEVERITY'),
+                ])
+                    // TODO Note: the backend does not appear to support the documented pagination queries, so we
+                    // need to limit the returned number of results client side
+                    .slice(0, limit)
+                    // We reverse here, because PF/Victory charts stack the bars from bottom->up
+                    .reverse();
```

Comment:
Two questions:
* Is `limit` for the number of policy categories? Looking from the outside in, it is not apparent that the widget has filtered the data in this way.
* To call `reverse` for Victory chart seems to couple data fetching with presentation implementation.

Maybe a question to discuss in team meeting how to balance convenience with coupling.

Not requesting a change, but suggesting to evaluate pro and con of an alternative:
* use specific hook to reduce boilerplate of `useEffect` with fetch function and 3 `useState` hooks
* call pure function to transform response in component that renders the chart (question where to declare)
* optional: reverse the return value also to make it visible in the component that renders the chart

Reply:
Initially I was aiming to use `limit` to pass to the pagination part of the API since we only want the top 5 results for the widget. However there were two problems with this: the server doesn't appear to support the pagination API, and more importantly, we need to do the sorting client side since the sort method is a bit more complex that what the backend can do.

I think it definitely makes sense to move the `limit` and `reverse` calls out of here since they are coupled to the presentation. Actually now that I think of it the sorting functionality should be moved out as well. The next iteration is going to allow the user to change the way sorting is applied, so I'll retain the server response in the hook state and run pure functions in the component for presentation changes.

Reply:
For a boilerplate reducing hook that wraps our fetch calls, I'm all for that idea. I know there was some discussion a while back regarding bringing in react-query, or pulling [this code](https://github.com/stackrox/infra/blob/master/ui/src/client/useApiQuery.ts) from stackrox/infra.

I was thinking I would refactor this as a follow up, since we'll want to have some team discussions on what approach works best for everyone and possibly draft up a design doc sometime after Saif gets back.